File: renderer/src/state/patch.ts
// TS twin of planner-core::patch::apply — entity-level ops on the projected plan.

import type { Plan, PatchOp } from "./types";

export function applyPatches(plan: Plan, ops: PatchOp[]): Plan {
  const next: Plan = {
    ...plan,
    factories: { ...plan.factories },
    groups: { ...plan.groups },
    ports: { ...plan.ports },
    edges: { ...plan.edges },
    nodeClaims: { ...plan.nodeClaims },
    routes: { ...plan.routes },
    junctions: { ...plan.junctions },
    proposals: { ...plan.proposals },
    switches: { ...plan.switches },
    styleGuides: { ...plan.styleGuides },
    meta: { ...plan.meta },
  };
  for (const op of ops) {
    const path = op.path.replace(/^\//, "");
    const slash = path.indexOf("/");
    if (slash < 0) continue;
    const collection = path.slice(0, slash);
    const key = path.slice(slash + 1);
    if (collection === "meta") {
      if (op.op !== "remove") (next.meta as unknown as Record<string, unknown>)[key] = op.value;
      continue;
    }
    const coll = (next as unknown as Record<string, Record<string, unknown>>)[collection];
    if (!coll) continue;
    if (op.op === "remove") delete coll[key];
    else coll[key] = op.value;
  }
  return next;
}

File: crates/planner-core/src/patch.rs
//! Entity-granular JSON patches. A `PatchBatch` is what every command returns and
//! what `state://patch` events carry (SDD §4). Paths address the projected-state
//! tree the renderer holds: `/factories/<id>`, `/groups/<id>`, `/ports/<id>`,
//! `/edges/<id>`, `/nodeClaims/<id>`, `/meta/...`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum PatchOp {
    Add { path: String, value: Value },
    Replace { path: String, value: Value },
    Remove { path: String },
}

impl PatchOp {
    pub fn path(&self) -> &str {
        match self {
            PatchOp::Add { path, .. }
            | PatchOp::Replace { path, .. }
            | PatchOp::Remove { path } => path,
        }
    }
}

pub type PatchBatch = Vec<PatchOp>;

/// Apply a batch to a JSON tree (two-segment paths: `/collection/id`).
/// Used by tests and by state reconstruction; the renderer has a TS twin.
pub fn apply(tree: &mut Value, batch: &[PatchOp]) -> Result<(), String> {
    for op in batch {
        let (coll, key) = split_path(op.path())?;
        let obj = tree
            .get_mut(coll)
            .and_then(Value::as_object_mut)
            .ok_or_else(|| format!("unknown collection {coll}"))?;
        match op {
            PatchOp::Add { value, .. } | PatchOp::Replace { value, .. } => {
                obj.insert(key.to_string(), value.clone());
            }
            PatchOp::Remove { .. } => {
                obj.remove(key);
            }
        }
    }
    Ok(())
}

fn split_path(path: &str) -> Result<(&str, &str), String> {
    let mut it = path.trim_start_matches('/').splitn(2, '/');
    match (it.next(), it.next()) {
        (Some(c), Some(k)) if !c.is_empty() && !k.is_empty() => Ok((c, k)),
        _ => Err(format!("bad patch path {path}")),
    }
}
